### PR TITLE
[Relation] Add MaterializedRelation

### PR DESCRIPTION
**Pull request by Tishj**
This PR adds a new Relation type.

### `MaterializedRelation`

This Relation represents a materialized data set, likely produced by a MaterializedQueryResult.
It contains a ColumnDataCollection and allows us to efficiently scan this with the existing Logical + PhysicalColumnDataScan.

### `ColumnDataRef`

To do this it introduces the ColumnDataRef and BoundColumnDataRef classes.
We don't want to make unnecessary copies of this (potentially giant) column data collection, so the ColumnDataRef only gets a reference to the collection.

Because TableRef's need to be Serializable though, we can't rely on this reference after deserialization.
If the ColumnDataRef is serialized to disk, it writes the entire column data collection to disk as well.
When the ColumnDataRef is deserialized, it creates a copy of the collection that it now owns.

As a result, the ColumnDataRef has an `optionally_owned_ptr`, and this trickles down into `BoundColumnDataRef`, `LogicalColumnDataScan` and `PhysicalColumnDataScan` (which already had this quality)

### `optionally_owned_ptr`
To encapsulate the scenario where a class contains this pattern:
```c++
//! The owned object (if any)
unique_ptr<T> owned_object;
//! The referenced object
optional_ptr<T> object;
```

We create a new class `optionally_owned_ptr` which makes sure that these constraints are respected:
- if `owned_object` is set, `object` must refer to it
- if the instance is moved, both pointers become null
- the instance can not be copied

`is_owned()` can be used to check whether the pointer is owning
Because the class is neither an `optional_ptr` or a `unique_ptr` it is intentionally not convertible to either.

### Usage

In the Python API, in the internal execute method (shared by query/sql and execute) we previously created a ValueRelation if the statement was not a SELECT, this is now replaced by a MaterializedRelation.

### Benchmark results:

These are timings measuring `fetchall()` from relations created with either `CALL` or `SELECT` statements.
The CALL statements utilize the MaterializedRelation.
```
call_repeat_row_10000000
timing: 4.085824966430664

call_repeat_row_2048
timing: 0.0007717609405517578

call_repeat_row_2500000
timing: 1.0177351236343384

call_repeat_row_50000
timing: 0.018980026245117188

select_repeat_row_10000000
timing: 4.076231598854065

select_repeat_row_2048
timing: 0.0007649660110473633

select_repeat_row_2500000
timing: 1.0249409675598145

select_repeat_row_50000
timing: 0.019537687301635742
```

### Future work

The plan is to extend the DuckDBPyRelation to store a MaterializedRelation internally when `DuckDBPyRelation.execute` is called, any future reference of the relation after that (by a replacement scan or by incrementally building on top of it) will make use of the MaterializedRelation.

### CODE REVIEW

**Comment by carlopi** [comment]
As always, impressive work.

Only note for posterity: this might make so that going through serialization and then deserialization we might start from 1 object somewhere in memory to having a few copies, potentially with degenerate cases like circular linked list that expand to infinite data.

Responsibility is on the users of optionally_owned_ptr to guarantee that data structure are acyclical.

File: src/include/duckdb/main/relation/materialized_relation.hpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/main/relation/value_relation.hpp
```
Comment (Mytherin) [inline src/include/duckdb/main/relation/materialized_relation.hpp:4]: nit

File: tools/pythonpkg/tests/fast/spark/test_spark_join.py [hunk 2]
```diff
@@ -45,15 +45,15 @@ class TestDataFrameJoin(object):
     def test_inner_join(self, dataframe_a, dataframe_b):
         df = dataframe_a.join(dataframe_b, dataframe_a.emp_dept_id == dataframe_b.dept_id, "inner")
         res = df.collect()
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/spark/test_spark_join.py:47]: Can this have a sort or order so this doesn't break in the future?

File: tools/pythonpkg/tests/fast/spark/test_spark_join.py [hunk 3]
```diff
@@ -80,42 +91,32 @@ def test_inner_join(self, dataframe_a, dataframe_b):
...
-                dept_name='IT',
-                dept_id=40,
-            ),
         ]
+        assert res == expected
 
     @pytest.mark.parametrize('how', ['outer', 'fullouter', 'full', 'full_outer'])
     def test_outer_join(self, dataframe_a, dataframe_b, how):
         df = dataframe_a.join(dataframe_b, dataframe_a.emp_dept_id == dataframe_b.dept_id, how)
         res1 = df.collect()
```
Comment (Mytherin) [inline tools/pythonpkg/tests/fast/spark/test_spark_join.py:110]: Can this have a sort or order?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great - some minor comments.

* Maybe we can also do some tests where we create a materialized relation and then query it.
* Perhaps also something like, we destroy the connection it came from and then query it in a different one.
* Maybe also some performance tests, e.g. how fast is the query of a materialized query result?

**Comment by Mytherin** [comment]
The benchmarks seem to use too much memory for the CI, could you perhaps reduce the load there?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**